File: 285-inorder-successor-in-bst/285-inorder-successor-in-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        TreeNode prev = null;
        
        while(root != null && root.val != p.val) {
            if (root.val < p.val) {
                root = root.right;
                continue;
            }
            
            prev = root;
            root = root.left;
        }
        
        if (root == null) return null;
        
        TreeNode next = root.right;
        if (next == null) return prev;
        
        while(next.left != null) next = next.left;
        
        return next;
    }
}

File: 285-inorder-successor-in-bst/285-inorder-successor-in-bst.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */

object Solution {
    def inorderSuccessor(root: TreeNode, p: TreeNode): TreeNode = {
        var current = root
        
        while(current != null && current.value != p.value) {
            var next: TreeNode = null
            
            if (current.value < p.value) {
                next = current.right
                
                var prev = next.left
                while(prev != null && prev.right != next) prev = prev.right
                
                if (prev != null) prev.right = null
                
            } else {
                next = current.left
                
                var prev = next
                while(prev != null && prev.right != null) prev = prev.right
                
                if (prev != null) prev.right = current
            }
            
            current = next
        }
        
        if (current == null) null
        else {
            var next = current.right
            var prev = if (next == null) null else next.left
            
            while(prev != null && prev != next) prev = prev.right
            
            if (prev == null) while(next != null && next.left != null) next = next.left
            
            next
        }
    }
}